### ISSUE TITLE
Curtain and Blind on my Tahoma. Only blind showing in Homekit (through homebridge)

### ISSUE THREAD

**Issue by github2302** [dubocr/homebridge-tahoma#29]
Hello, thanks to all who have made the Homebridge and Tahoma code which I have running on my Raspberry Pi along with Sonos plugin too.

I have curtains in the bedroom and roman blinds in the living room which are both powered by Somfy Tahoma, and I can raise/lower and open/close using the Somfy Tahoma app. I've added the Tahoma plugin to my homebridge and immediately can see the living room blinds which work perfectly with Siri. **However the bedroom curtains do not appear in the Home app on my iPhone.**

Is there a limit to only one curtain/blind device? Or is there a known problem with curtains? Or perhaps I've done something wrong?

Very grateful for any help! Thank you!

**Comment by dubocr** [comment]
Hello,

Can you please log here to get support : http://home.dubocr.fr/tools/setup

Thanks

**Comment by github2302** [comment]
Hello, I have done just now. Will that automatically send you what you need, or do I need to copy and paste / export something?  Thanks.

**Comment by dubocr** [comment]
That's good, can you add the following js file into the homebridge-tahoma 'accessories' directory and restart homebridge : 
[Curtain.js.zip](https://github.com/dubocr/homebridge-tahoma/files/2066208/Curtain.js.zip)

**Comment by github2302** [comment]
I don't see an accessories folder to drop this into, can you point me in the right direction? I've looked in ~./homebridge and /usr/local/bin

Also, is there a command to stop homebridge without restarting the pi?

Thanks

**Comment by github2302** [comment]
Also, I now have this after a restart of the Pi, but no success. Does this look right?

pi@raspberrypi:/usr/local/lib/node_modules/homebridge-tahoma/accessories $ more Curtain.js 
module.exports = function(homebridge, abstractAccessory, api) {
    return require('./Awning.js')(homebridge, abstractAccessory, api);
}

pi@raspberrypi:/usr/local/lib/node_modules/homebridge-tahoma/accessories $

**Comment by Shana611** [comment]
I have the same problem. I've already added the Cassette Skylight blinds to TaHoma that's working via Tahoma app. But it doesn't show on the AppleHomeKit. Pls can you help me. Thanks

**Comment by dubocr** [comment]
Hi,

Your Curtains are RTS devices so they don't send back their status and only 3 commands are available.
In HomeKit, the corresponding compatibility behaviour is :
- Open (100%) : open the curtains
- Close (0%) : close the curtains
- Any intermediate values (1 to 99%) : set the curtains in "my" position (25% in your case !?)

When setting intermediate values (ex. 50%), Home normally display the corresponding value at the end of the motion (so 50%).
Can you confirm me that the app is displaying it as fully closed when you select an intermediate value ?

`sudo systemctl restart homebridge` will restart homebridge if you installed it as service.

**Comment by Shana611** [comment]
Thank you so much for your quick reply. Window blinds show & work perfectly on AppleHomwkit but the problem is, I have two Casette Roller Blinds to connected to the TaHoma But it doesn’t appear on HomeKit. I done the systmctl restart but no luck.

**Comment by github2302** [comment]
Confirming that as you said, choosing any intermediate percentage in Home app or by Siri takes the blinds to “My” position. In this position, the app shows as “Opening” for a long while and then eventually shows either open, closed or sometimes open XX% (where XX is the chosen percentage). It seems to be different every time I go into intermediate!


On 5 June 2018 at 20:34:35, Shana611 (notifications@github.com<mailto:notifications@github.com>) wrote:

Thank you so much for your quick reply. Window blinds show & work perfectly on AppleHomwkit but the problem is, I have two Casette Roller Blinds to connected to the TaHoma But it doesn’t appear on HomeKit. I done the systmctl restart but no luck.

**Comment by dubocr** [comment]
@Shana611 : can you log [there](http://home.dubocr.fr/tools/setup) to let me read your config. Thanks
@github2302 : can you send me homebridge logs `sudo systemctl status homebridge` after setting your blinds in intermediate position with Siri (after everything is done)

**Comment by Shana611** [comment]
I logged in as you said

**Comment by dubocr** [comment]
@Shana611 : add this file to the plugin 'accessories' directory then restart : 
[Generic.js.zip](https://github.com/dubocr/homebridge-tahoma/files/2073887/Generic.js.zip)

**Comment by Shana611** [comment]
Ok I will & I updat you the new status. Thank you so much

**Comment by Shana611** [comment]
Do I need to edit anything in (config.json

**Comment by dubocr** [comment]
Nop

**Comment by Shana611** [comment]
Hi I manged to get it work. The problem was The commands on Awning.js should change it to Close Open  to Up Down. But 50% problem still there. Anyway Thank you soooo much again for your help.

**Comment by dubocr** [comment]
50% is the "defaultPosition" value (could be changed in config file) at homebridge startup. When you change the state of your blinds, its remains to the selected value except if homebridge is restarted. The choice of 50% is to allow you to open/close your blinds even after a homebridge restart.

Here is the new file with the correction you already done (up/down) [Awning.js.zip](https://github.com/dubocr/homebridge-tahoma/files/2082006/Awning.js.zip)

**Comment by github2302** [comment]
Hi @dubocr, sorry for the delay replying as I've been travelling. I restarted Homebridge, and now all seems happy! Siri/app opens/closes correctly and displays correctly in the home app. Siri/app moves the units to the 'My' position when asked for any % level. Shows in the app as at that % level.

[2018-6-12 09:44:41] [Tahoma] [Living Room] close[]
[2018-6-12 09:45:14] [Tahoma] [Living Room] close COMPLETED
[2018-6-12 09:45:18] [Tahoma] [Living Room] my[]
[2018-6-12 09:45:50] [Tahoma] [Living Room] my COMPLETED

Do you think there will ever be a way for the Somfy blinds to go to the percentage set by Siri, or is this a limitation of the Somfy blinds?

Thank you again.

**Comment by Marderren** [comment]
Hello, sorry for hijacking this thread. But I have two questions related to above.

I have added my Awning to TaHoma and running Homebridge on Raspnerry Pi.

1. For some reason the In / Out states seems to have bene reversed. Meaning when I roll out the Awning completely my Home App says “0% / Rolled in”and when I roll the awning in it says “100% / Rolled out”. Is there any way to change this behaviour? I have tried to roll out the awning out and then restart Homebridge but the issue persists.

2. Also how do I change the 50% start value? It seems to cause a little bit of trouble for my setup that the starting value is 50% and sometimes the Home App won’t control the awning before I have used the manual Somfy remote first.. What confit file should be changed and to what?

Thank.

**Comment by dubocr** [comment]
Hi, can you log there (http://home.dubocr.fr/tools/setup) to give me your config and tel me the name of Awning that are reversed. Tks

**Comment by Marderren** [comment]
Sure! I am home again Saturday and will take a look at it then :)

**Comment by Marderren** [comment]
@dubocr I have just sent the config through the above link now. The name of the awning is the only device in there, but just in case it's called "Markise" or "Awning > UpDownHorizontalAwning"

**Comment by Marderren** [comment]
Hi @dubocr did you receive my config file? Do you know what is causing the reversed setting in Apple Home? Would it be something I can fix myself by any chance? Any help would be much appreciated :)

**Comment by dubocr** [comment]
Hi,

Can you try to replace Anwing file in accessories by this one.

[Awning.js.zip](https://github.com/dubocr/homebridge-tahoma/files/2236043/Awning.js.zip)

Thks

**Comment by Marderren** [comment]
Thanks, but I'm not sure where to put the Awning file to be honest :-)

In my Homebridge folder on my Raspberry Pi I only see 2 folders besides my "config.json" file.

Folder: "accessories" only contains a "cashedAccessories" file

Folder: "persist" only contains "AccessoryInfo.B827EBFA00B4.json" and IdentifierCache.B827EBFA00B4.json

**Comment by dubocr** [comment]
Ok so try to uninstall then reinstall tahoma plugin.

**Comment by Marderren** [comment]
I think my problem is that I am running Homebridge on Home Assistant (Hass.io). I am not sure where they keep the actual JS files. To install Tahoma-plugin on Hass.io I simply add 
 "plugins": [
    "homebridge-tahoma"]
in the Hass.io Homebridge config file before starting the service.

Maybe I have to try to install Homebridge separately outside Home Assistant. Will let you know how it goes :)

**Comment by Marderren** [comment]
Hi @dubocr I've now tried with a completely new Pi and fresh installation, installed Homebridge the "old fashioned" way without Hass.io

I realized the accessories folder I was looking at yesterday was the /home/pi/.homebridge/accessories
and not the /usr/lib/node_modules/homebridge-tahoma :-) 

After installing homebridge-tahoma plugin this morning I saw the Awning.js version in tahoma plugin foldwer was now the same as you sent me. 
But unfortunately the awning is still reversed:
100%/Open = Awning physically rolled in/closed
0%/Closed in App = Awning physically rolled out/open

Any ideas?

**Comment by Marderren** [comment]
Hello everyone! Just wanted to say that I have kind of fixed this issue myself. I basically just took a wild guess and opened the Awning.js file from @dubocr in Note++ and changed 100 to 0 and 0 to 100 two places like in below example:

_value = ['Screen', 'Awning'].includes(this.device.uiClass) ? (100-value) : value; // Reverse Screen devices
switch(value) {
	case 100: command = new Command('up'); break;
	case 0: command = new Command('down'); break;
	default: command = new Command('my'); break;_

to

_value = ['Screen', 'Awning'].includes(this.device.uiClass) ? (100-value) : value; // Reverse Screen devices
switch(value) {
	case 0: command = new Command('up'); break;
	case 100: command = new Command('down'); break;
	default: command = new Command('my'); break;_

My Somfy RTS Awning now works as expected and I can say "Hey Siri set Awning to 100%" and the awning will roll out. 

I am a Github n00b, so I don't know how to change this in the official package or anything, but perhaps someone can help with that. (Note: I am not sure if it was actually only MY awning that was reversed or if this was a general issue with the RTS awnings).

**Comment by dubocr** [comment]
Hi,

Seems there is something strange as your Awning are not concerned by this part of code (that's the reason my fix has no effect).
I tried to normalize this recurrent issue by making 4 Anwing categories :

- UpDownHorizontal (use deploy/undeploy)
- PositionableHorizontal (use setDeployment)
- UpDown/RTS one (use up/down)
- Others (use setClosure)

Your Awning must use the first one (UpDownHorizontal).
Could you please try this new file, tell me if the behavior is the good one and check if the called method is 'deploy' or 'undeploy' in homebridge logs `sudo systemctl status homebridge`
[Awning.js.zip](https://github.com/dubocr/homebridge-tahoma/files/2252983/Awning.js.zip)

Your Tahoma account is the following one ? m******.o******[AT]gmail.com

Thanks

**Comment by Marderren** [comment]
Ok just tried your latest file. Clicking/touching the Awning in Apple Home App (the icon that has 50% on startup) now does not work at all as one would expect. 

- First click/touch would deploy awning to the MySetting position (Homebridge cmd says "deploy[]"

- Second click would undeploy/roll back the awning ("undeploy[]")

- Third click would deploy the awning to 50%, then 1 minute later it would complete and go to 100% Homebridge log says:
deploy[]
my TIME_OUT_ON_COMMAND_PROGRESS
deploy COMPLETED

BUT! No matter what happens in all of the above scenarios the Apple Home App icon keep saying 50% now. 

If I hold my touch on the icon to bring up the slider-menu and then select 100% the awning will deploy/roll out completely, but it still says 50% in the App.

If I hold my touch on the icon to bring up the menu and select 0% the awning will first roll in to MyPosition, then after 30-60seconds it would continue to roll back in completely. Homebridge log says:
my[]
undeploy[]

----

Please note that the file mentioned in previous post where I manually changed the 100 to 0, and 0 to 100 values works perfectly for me :-) 

P.s. yes that is my Tahoma account :)

**Comment by Marderren** [comment]
Just tried with my manually edited file that works for me. With this file Homebridge log says:
open[] when I undeploy/close the awning
close[] when I deploy/open the awning 
my[] when I input something between 1% and 99%

(Just as one would expect since I basically just reverted the numbers).

**Comment by dubocr** [comment]
Yes, as you said, the commands are open/close and in the section of code you modified it's up/down. So my and your update is not responsible of the good behavior.
This new file will fix the always 50% behavior : 
[Awning.js.zip](https://github.com/dubocr/homebridge-tahoma/files/2253301/Awning.js.zip)

Here is the expected behavior :

Start : Awning at 50%

1st touch : Homekit closing... => undeploy
After completion : Homekit closed

2nd touch : Homekit opening... => deploy
After completion : Homekit opened

Slider between 1 and 99% : Homekit opening/closing... => my
After completion : Homekit X%

Is the previous behavior the expected one for you ? Can you tell me if the attached file executes it well.

Thanks

**Comment by Marderren** [comment]
Thank you! Your latest file works as expected :) 

homebridge[491]: [2018-8-2 14:26:59] [Tahoma] [Markise] undeploy[]
homebridge[491]: [2018-8-2 14:27:31] [Tahoma] [Markise] undeploy COMPLETED
homebridge[491]: [2018-8-2 14:27:36] [Tahoma] [Markise] deploy[]
homebridge[491]: [2018-8-2 14:28:09] [Tahoma] [Markise] deploy COMPLETED
homebridge[491]: [2018-8-2 14:28:16] [Tahoma] [Markise] my[]
homebridge[491]: [2018-8-2 14:28:49] [Tahoma] [Markise] my COMPLETED

**Comment by azreik198** [comment]
Dear how i can import scenarios from tahoma to homebridge
any help

**Comment by dubocr** [comment]
Hi,
Scenario are not supported by homebridge but you can add Tahoma scenario as on/off switch by adding the parameter "exposeScenarios": true in your config :
